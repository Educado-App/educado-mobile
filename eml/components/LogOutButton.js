//Navigation til Login Screen

//Feather icons

//State på at man har logget ud

//bevar state til næste åbning af app

//importer state i login form

import React, {useState} from 'react';
import {StyleSheet, Image, View, Dimensions, TouchableOpacity, Text, Alert, Pressable} from "react-native";
import {useNavigation} from "@react-navigation/native";
import { Feather } from '@expo/vector-icons';

export default function LogOutButton(props) {

    const navigation = useNavigation();

    const logoutAlert = () =>
        Alert.alert(
            "Logout",
            "Are you sure?",
            [
                {
                    text: "No",
                    onPress: () => console.log("No Pressed"),
                    style: "cancel"
                },
                { text: "Yes", onPress: () => navigation.navigate('Login')}
            ]
        );

    return (
        <View style ={styles.container}>
            <TouchableOpacity style={styles.formButton} onPress={logoutAlert}>
                <Feather name="log-out" size={36}  style={styles.tinyLogo} />
            </TouchableOpacity>
        </View>
    );
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        justifyContent : 'flex-start'
    },
 /*   button: {
        backgroundColor : 'rgba(123,104,238,0.8)',
        height : 55,
        alignItems : 'center',
        justifyContent : 'center',
        borderRadius: 35,
        marginHorizontal : 20,
        marginVertical: 10,
        borderWidth: 1,
        borderColor: 'white'
    },*/
    buttonText:{
        fontSize: 20,
        fontWeight: '600',
        color: 'black',
        letterSpacing: 0.5
    },
    bottomContainer:{
        marginVertical: '150%',
        justifyContent: 'center',
        height: '25%',
    },
    formButton: {
        backgroundColor : 'white',
        width : 75,
        height : 55,
        alignItems : 'center',
        justifyContent : 'center',
        borderRadius: 15,
        marginHorizontal : 325,
        marginVertical: 10,
        borderWidth: 1,
        borderColor: 'black',
        shadowColor: "#000",
        shadowOffset: {
            width: 0,
            height: 4,
        },
        shadowOpacity: 0.7,
        shadowRadius: 3.84,
        elevation: 5,
    },
    formInputContainer:{
        marginBottom: 70
    },
    tinyLogo: {
        width: 50,
        height: 50,
    }
});
