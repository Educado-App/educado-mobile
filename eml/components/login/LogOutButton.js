import {StyleSheet, View, TouchableOpacity, Alert} from "react-native";
import {useNavigation} from "@react-navigation/native";
import { Feather } from '@expo/vector-icons';
import AsyncStorage from '@react-native-async-storage/async-storage'

const LOGIN_TOKEN = '@loginToken';
const USER_INFO = '@userInfo'

export default function LogOutButton() {

    const navigation = useNavigation();

    const logoutAlert = () =>
        Alert.alert(
            "Logout",
            "Are you sure?",
            [
                {
                    text: "No",
                    onPress: () => console.log("No Pressed"),
                    style: "cancel"
                },
                { text: "Yes", onPress: () =>{

                    try {
                        AsyncStorage.multiRemove([LOGIN_TOKEN, USER_INFO]).then(r => {
                            console.log("User Logged out successfully!");
                            navigation.navigate('LoginStack');
                        });
                    }
                    catch (e){
                        console.log(e);
                    }

                }}
            ]
        );

    return (
        <View style ={styles.container}>

            <TouchableOpacity style={styles.formButton} onPress={logoutAlert}>
                <Feather name="log-out" size={36}  style={styles.tinyLogo} />
            </TouchableOpacity>

            <View style={styles.bottomContainer}>
                <View style={styles.formInputContainer}>
                    <TouchableOpacity onPress={() => navigation.navigate('Login')}>
                        <View style={styles.formButton}>
                            <Feather name="log-out" size={36}  style={styles.menu} />
                        </View>
                    </TouchableOpacity>
                </View>
            </View>

        </View>
    );
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        justifyContent : 'flex-start'
    },
    buttonText:{
        fontSize: 20,
        fontWeight: '600',
        color: 'black',
        letterSpacing: 0.5
    },
    bottomContainer:{
        marginVertical: '150%',
        justifyContent: 'center',
        height: '25%',
    },
    formButton: {
        backgroundColor : 'white',
        width : 75,
        height : 55,
        alignItems : 'center',
        justifyContent : 'center',
        borderRadius: 15,
        marginHorizontal : 325,
        marginVertical: 10,
        borderWidth: 1,
        borderColor: 'black',
        shadowColor: "#000",
        shadowOffset: {
            width: 0,
            height: 4,
        },
        shadowOpacity: 0.7,
        shadowRadius: 3.84,
        elevation: 5,
    },
    formInputContainer:{
        marginBottom: 70
    },
    tinyLogo: {
        width: 50,
        height: 50,
    }
});
