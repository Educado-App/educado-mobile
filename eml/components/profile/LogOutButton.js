import { View, TouchableOpacity, Alert } from 'react-native';
import { useNavigation } from '@react-navigation/native';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { MaterialCommunityIcons } from '@expo/vector-icons';
import React from 'react';
import Text from '../general/Text';
import tailwindConfig from '../../tailwind.config';
import PropTypes from 'prop-types';


const LOGIN_TOKEN = '@loginToken';
const USER_INFO = '@userInfo';
const tailwindColors = tailwindConfig.theme.colors;
const STUDENT_INFO = '@studentInfo';

export default function LogOutButton(props) {
	LogOutButton.propTypes = {
		testID: PropTypes.string,
	};

	const navigation = useNavigation();

	async function logOut() {
		try {
			await AsyncStorage.removeItem(LOGIN_TOKEN);
			await AsyncStorage.removeItem(USER_INFO);
			await AsyncStorage.removeItem(STUDENT_INFO);

			navigation.navigate('LoginStack');
		} catch (e) {
			console.log(e);
		}
	}

	const logoutAlert = () =>
		Alert.alert('Sair', 'Tem certeza que deseja sair?', [
			{
				text: 'NÃ£o',
				onPress: () => console.log('No Pressed'),
				style: 'cancel'
			},
			{ text: 'Sim', onPress: logOut }
		]);

	return (
		<View className="items-center flex-1 mt-[12%]">
			<TouchableOpacity onPress={logoutAlert}>
				<View className='items-center flex flex-row'>
					<MaterialCommunityIcons
						name="logout"
						size={30}
						color={tailwindColors.error}
						testID={props.testID}
					/>
					<Text className="text-error text-center underline font-sans-bold">Sair</Text>
				</View> 
        
			</TouchableOpacity>
		</View>
	);
}