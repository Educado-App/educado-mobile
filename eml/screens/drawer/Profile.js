import {View, StyleSheet} from 'react-native';
import LogOutButton from '../../components/profile/LogOutButton';
import {Feather} from '@expo/vector-icons';
import React from 'react';
import Text from '../../components/general/Text';

export default function ProfileScreen() {
  return (
    <View>
      <View>
        {/* Profile Screen */}
        <Text style={styles.textLogoContainer}>Tela de perfil</Text>
        <Feather name="user" size={48} style={styles.tinyLogo}/>
      </View>
      <LogOutButton></LogOutButton>
    </View>

  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent : 'flex-start'
  },

  tinyLogo: {
    marginVertical: 0,
    marginHorizontal: 190,
    width: 50,
    height: 50,
    alignItems : 'center',
    justifyContent : 'center',
  },
  textLogoContainer: {
    marginHorizontal : '26%',
    marginVertical: '15%',
    fontSize: 35,
    fontWeight: '400',
    color: 'green',
    letterSpacing: 0.5,
    height: 50,
    width: 1000,
    justifyContent: 'center'
  }
});
