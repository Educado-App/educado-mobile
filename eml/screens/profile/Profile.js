import React, { useEffect, useState } from 'react';
import {
	View,
	SafeAreaView,
	ScrollView,
} from 'react-native';
import LogOutButton from '../../components/profile/LogOutButton';
import ProfileNavigationButton from '../../components/profile/ProfileNavigationButton.js';
import UserInfo from '../../components/profile/UserInfo';
import { useNavigation } from '@react-navigation/native';
import { getUserInfo } from '../../services/StorageService';
import errorSwitch from '../../components/general/errorSwitch';
import ShowAlert from '../../components/general/ShowAlert';
import { getStudentInfo } from '../../services/StorageService';
import ProfileStatsBox from '../../components/profile/ProfileStatsBox';

/**
 * Profile screen
 * @returns {React.Element} Component for the profile screen
 */
export default function ProfileComponent() {
	const [firstName, setFirstName] = useState('');
	const [lastName, setLastName] = useState('');
	const [email, setEmail] = useState('');
	const navigation = useNavigation();
	const [studentLevel, setStudentLevel] = useState(0);
	const [levelProgress, setLevelProgress] = useState(0);
	const [totalPoints, setTotalPoints] = useState(0);

	useEffect(() => {
		const getInfo = navigation.addListener('focus', () => {
			getProfile();
		});
		return getInfo;
	}, [navigation]);

	const getLevelProgress = (student) => {
		const pointsForPreviousLevel = (student.level - 1) * 100;
		const pointsForNextLevel = student.level * 100;

		return ((student.points - pointsForPreviousLevel)/(pointsForNextLevel - pointsForPreviousLevel)) * 100;
	};

	/**
  * Fetches the user's profile from local storage
  */ 
	const getProfile = async () => {
		try {
			const fetchedProfile = await getUserInfo();
			const fetchedStudent = await getStudentInfo();
			if (fetchedProfile !== null) {
				setFirstName(fetchedProfile.firstName);
				setLastName(fetchedProfile.lastName);
				setEmail(fetchedProfile.email);
			} else if (fetchedStudent !== null) {
				setStudentLevel(fetchedStudent.level);
				setTotalPoints(fetchedStudent.points);
				setLevelProgress(getLevelProgress(fetchedStudent));
			}
		} catch (error) {
			ShowAlert(errorSwitch(error));
		}
	};

	useEffect(() => {
		getProfile();
	}, []);

	const fetchStudentProfile = async () => {
		const studentInfo = await getStudentInfo();
		setStudentLevel(studentInfo.level);
		setTotalPoints(studentInfo.points);
		setLevelProgress(getLevelProgress(studentInfo));
	};
  
	useEffect(() => {
		fetchStudentProfile();
	}, []);

	return (
		<SafeAreaView className='bg-secondary'>
			<ScrollView className='flex flex-col'>
				<View className="flex-1 justify-start pt-[20%] h-screen">
					<UserInfo firstName={firstName} lastName={lastName} email={email} points={totalPoints}></UserInfo>
					<ProfileStatsBox studentLevel={studentLevel} levelProgress={levelProgress} />
					<ProfileNavigationButton label='Editar perfil' testId={'editProfileNav'} onPress={() => navigation.navigate('EditProfile')}></ProfileNavigationButton>
					<ProfileNavigationButton label='Certificados' onPress={() => navigation.navigate('CertificateStack')}></ProfileNavigationButton>
					<ProfileNavigationButton label='Download'></ProfileNavigationButton>
					<View className='flex flex-row pb-4'>
						<LogOutButton testID='logoutBtn'></LogOutButton>
					</View>
				</View>
			</ScrollView>
		</SafeAreaView>
	);
}