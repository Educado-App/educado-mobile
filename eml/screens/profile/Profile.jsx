import React, { useEffect, useState, useCallback } from 'react';
import ToastNotification from '../../components/general/ToastNotification';
import AsyncStorage from '@react-native-async-storage/async-storage';
import {
	View,
	ScrollView,
} from 'react-native';
import LogOutButton from '../../components/profile/LogOutButton';
import ProfileNavigationButton from '../../components/profile/ProfileNavigationButton.js';

import UserInfo from '../../components/profile/UserInfo';
import { useNavigation } from '@react-navigation/native';
import { getUserInfo } from '../../services/StorageService';
import errorSwitch from '../../components/general/errorSwitch';
import ShowAlert from '../../components/general/ShowAlert';
import { getStudentInfo } from '../../services/StorageService';
import ProfileStatsBox from '../../components/profile/ProfileStatsBox';
import { useFocusEffect } from '@react-navigation/native';

/**
 * Profile screen
 * @returns {React.Element} Component for the profile screen
 */
export default function ProfileComponent() {
	const [firstName, setFirstName] = useState('');
	const [lastName, setLastName] = useState('');
	const [email, setEmail] = useState('');
	const [photo, setPhoto] = useState('');
	const navigation = useNavigation();
	const [studentLevel, setStudentLevel] = useState(0);
	const [levelProgress, setLevelProgress] = useState(0);
	const [totalPoints, setTotalPoints] = useState(0);

	useEffect(() => {
		const getInfo = navigation.addListener('focus', () => {
			getProfile();
		});
		return getInfo;
	}, [navigation]);

	const getLevelProgress = (student) => {
		const pointsForPreviousLevel = (student.level - 1) * 100;
		const pointsForNextLevel = student.level * 100;

		return ((student.points - pointsForPreviousLevel) / (pointsForNextLevel - pointsForPreviousLevel)) * 100;
	};

	/**
  * Fetches the user's profile from local storage
  */
	const getProfile = async () => {
		try {
			const fetchedProfile = await getUserInfo();
			const fetchedStudent = await getStudentInfo();
			if (fetchedProfile !== null) {
				setFirstName(fetchedProfile.firstName);
				setLastName(fetchedProfile.lastName);
				setEmail(fetchedProfile.email);
				if (fetchedStudent !== null) setPhoto(fetchedStudent.photo);
			} else if (fetchedStudent !== null) {
				setStudentLevel(fetchedStudent.level);
				setTotalPoints(fetchedStudent.points);
				setLevelProgress(getLevelProgress(fetchedStudent));
			}
		} catch (error) {
			ShowAlert(errorSwitch(error));
		}
	};

	useFocusEffect(
		useCallback(() => {
			console.log('Profile screen focused');
			const runAsyncFunction = async () => {
				try {
					// Load profile data and check for password reset status
					await getProfile();
					await fetchStudentProfile();
					await checkPasswordReset();
				} catch (error) {
					console.error('Error fetching profile:', error);
				}
			};
	
			// Run the async function when the screen is focused
			runAsyncFunction();
		}, [])
	);


	const fetchStudentProfile = async () => {
		const studentInfo = await getStudentInfo();
		setStudentLevel(studentInfo.level);
		setTotalPoints(studentInfo.points);
		setLevelProgress(getLevelProgress(studentInfo));
	};

	const checkPasswordReset = async () => {
		try {
			console.log('Checking password reset');
			if (await AsyncStorage.getItem('passwordUpdated') == 'true') {
				ToastNotification('success', 'Senha alterada com sucesso');
				await AsyncStorage.setItem('passwordUpdated', 'false');
				return;
			}
		} catch (e) {
			console.error(e);
		}
	};

	return (
		<ScrollView className='flex flex-col'>
			<View className="flex-1 justify-start pt-[20%] h-screen">
				<UserInfo firstName={firstName} lastName={lastName} email={email} points={totalPoints} photo={photo}></UserInfo>
				<ProfileStatsBox studentLevel={studentLevel} levelProgress={levelProgress} />
				<ProfileNavigationButton label='Editar perfil' testId={'editProfileNav'} onPress={() => navigation.navigate('EditProfile')}></ProfileNavigationButton>
				<ProfileNavigationButton label='Alterar senha' onPress={() => navigation.navigate('EditPassword')}></ProfileNavigationButton>
				<ProfileNavigationButton label='Certificados' onPress={() => navigation.navigate('Certificate')}></ProfileNavigationButton>
				<ProfileNavigationButton label='Download' onPress={() => navigation.navigate('Download')}></ProfileNavigationButton>
				
				{/* The certificate page is created and works, it is only commented out to get it approved on play store
					<ProfileNavigationButton label='Certificados' onPress={() => navigation.navigate('CertificateStack')}></ProfileNavigationButton>*/}
				{/* Download page is not implemented yet. However, download works and can be accessed on home page when offline
				<ProfileNavigationButton label='Download'></ProfileNavigationButton>*/}
				<View className='flex flex-row pb-4'>
					<LogOutButton testID='logoutBtn'></LogOutButton>
				</View>
			</View>
		</ScrollView>
	);
}