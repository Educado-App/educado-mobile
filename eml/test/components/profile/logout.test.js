import AsyncStorage from '@react-native-async-storage/async-storage';
import React from 'react';
import renderer from 'react-test-renderer';
import Profile from '../../../screens/profile/Profile';
const { fireEvent, render } = require('@testing-library/react-native');
import LogOutButton from '../../../components/profile/LogOutButton';

jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({
    navigate: jest.fn(),
  }),
}));

describe('LogOutButton render', () => {
  it('should render correctly', () => {
    const logoutButton = renderer.create(<LogOutButton />);
    expect(logoutButton).toMatchSnapshot();
  });
});


describe('Logout', () => {
  it('should delete token when logging out', async () => {
    let profile;

    const mockToken = 'testToken';
    // Set token
    AsyncStorage.setItem('@loginToken', mockToken).then(async () => {
      // Check if token is set
      await AsyncStorage.getItem('@loginToken').expect(mockToken);
      profile = render(<Profile />);

      // Check if token is deleted when logout button pressed
      const LogOutButton = profile.root.findByProps({ testID: 'logoutBtn' });

      fireEvent.press(LogOutButton);

      // Check if token is null after logout
      const updatedToken = await AsyncStorage.getItem('@loginToken');
      expect(updatedToken).toBeNull();
    });
  });

  it('should delete userInfo when logging out', async () => {
    let profile;

    const mockUserInfo = {
      id: '123',
      firstName: 'Is',
      lastName: 'User',
      email: 'is@user.com',
    };

    // Set user info
    AsyncStorage.setItem('@userInfo', mockUserInfo).then(async () => {
      // Check if user info is set
      await AsyncStorage.getItem('@userInfo').expect(mockUserInfo);
      profile = render(<Profile />);

      // Press logout button
      const LogOutButton = profile.root.findByProps({ testID: 'logoutBtn' });

      fireEvent.press(LogOutButton);

      // Check if user info is null after logout
      const updatedUserInfo = await AsyncStorage.getItem('@userInfo');
      expect(updatedUserInfo).toBeNull();
    });
  });

  it('should navigate to login screen when logging out', () => {
    let profile;

    const mockToken = 'testToken';
    // Set token
    AsyncStorage.setItem('@loginToken', mockToken).then(async () => {
      // Check if token is set
      await AsyncStorage.getItem('@loginToken').expect(mockToken);
      profile = render(<Profile />);

      const LogOutButton = profile.root.findByProps({ testID: 'logoutBtn' });

      fireEvent.press(LogOutButton);

      // Check if navigation is called after logout
      expect(useNavigation().navigate).toHaveBeenCalled();
    });
  });
});