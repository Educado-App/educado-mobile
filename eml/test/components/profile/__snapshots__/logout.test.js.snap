// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogOutButton render should render correctly 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
      },
      {
        "flexBasis": "0%",
        "flexGrow": 1,
        "flexShrink": 1,
      },
      {
        "marginTop": "12%",
      },
    ]
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        [
          {
            "alignItems": "center",
          },
          {
            "display": "flex",
          },
          {
            "flexDirection": "row",
          },
        ]
      }
    >
      <Text />
      <Text
        style={
          [
            {
              "fontFamily": "Montserrat-Regular",
            },
            {
              "color": "#383838",
            },
            {
              "fontSize": 16,
            },
            [
              {
                "color": "#FF4949",
              },
              {
                "textAlign": "center",
              },
              {
                "textDecorationLine": "underline",
              },
              {
                "fontFamily": "Montserrat-Bold",
              },
            ],
          ]
        }
      >
        Sair
      </Text>
    </View>
  </View>
</View>
`;
